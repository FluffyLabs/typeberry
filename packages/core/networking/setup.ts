import QuicLogger, { formatting, LogLevel, StreamHandler } from "@matrixai/logger";
import { events, QUICClient, type QUICConnection, QUICServer, QUICSocket } from "@matrixai/quic";
import { BytesBlob } from "@typeberry/bytes";
import type { Ed25519Pair } from "@typeberry/crypto/ed25519.js";
import { Level, Logger } from "@typeberry/logger";
import {
  altNameRaw,
  certToPEM,
  ed25519AsJsonWebKeyPair,
  generateCertificate,
  type PeerInfo,
  privateKeyToPEM,
} from "./certificate.js";
import { getQuicClientCrypto, getQuicServerCrypto } from "./crypto.js";
import type { DialOptions } from "./network.js";
import { peerVerification } from "./peer-verification.js";
import { type PeerAddress, PeersManagement } from "./peers.js";
import { QuicNetwork } from "./quic-network.js";
import { QuicPeer } from "./quic-peer.js";
import { addEventListener } from "./quic-utils.js";

const logger = Logger.new(import.meta.filename, "net");

/** Networking server part options. */
export type Options = {
  /** Peer's ed25519key. */
  key: Ed25519Pair;
  /** Host interface to bound to. */
  host: string;
  /** Port to listen on. Use `0` for random. */
  port: number;
  /** Supported ALPNs (protocols) both for the inbound and outbound connections. */
  protocols: string[];
};

enum CloseReason {
  PeerIdMismatch = 0,
  DuplicateConnection = 1,
  ConnectionFromOurself = 2,
}

export class Quic {
  /** Setup QUIC socket and start listening for connections. */
  static async setup({ host, port, protocols, key }: Options): Promise<QuicNetwork> {
    const quicLoggerLvl = logger.getLevel() > Level.TRACE ? LogLevel.WARN : LogLevel.DEBUG;
    const quicLogger = new QuicLogger("quic", quicLoggerLvl, [
      new StreamHandler(formatting.format`${formatting.level}:${formatting.keys}:${formatting.msg}`),
    ]);

    // Load keypair
    const keyPair = ed25519AsJsonWebKeyPair(key);
    const privKeyPEM = await privateKeyToPEM(keyPair);
    const cert = await generateCertificate({
      certId: BytesBlob.blobFromString("QUIC Networking"),
      subjectKeyPair: keyPair,
      issuerKeyPair: keyPair,
    });

    const lastConnectedPeer = peerVerification();

    // QUICConfig
    const config = {
      keepAliveIntervalTime: 3000,
      maxIdleTimeout: 6000,
      applicationProtos: protocols,
      cert: certToPEM(cert),
      key: privKeyPEM,
      verifyPeer: true,
      verifyCallback: lastConnectedPeer.verifyCallback,
    };

    logger.info`🆔 Peer id: ** ${altNameRaw(key.pubKey)}@${host}:${port} ** (pubkey: ${key.pubKey})`;
    // Shared injected UDP socket
    const socket = new QUICSocket({
      logger: quicLogger.getChild("socket"),
    });

    // Start server on the socket.
    const server = new QUICServer({
      socket,
      config,
      crypto: getQuicServerCrypto(key),
      logger: quicLogger.getChild("server"),
    });

    // peer management
    const peers = new PeersManagement<QuicPeer>();

    // basic error handling
    addEventListener(server, events.EventQUICServerError, (error) => logger.error`🛜  Server error: ${error}`);
    addEventListener(server, events.EventQUICServerClose, (ev) => logger.error`🛜  Server stopped: ${ev}`);

    // handling incoming session
    addEventListener(server, events.EventQUICServerConnection, async (ev) => {
      const conn = ev.detail;
      if (lastConnectedPeer.info === null) {
        await conn.stop();
        return;
      }

      if (lastConnectedPeer.info.key.isEqualTo(key.pubKey)) {
        logger.log`🛜 Rejecting connection from ourself from ${conn.remoteHost}:${conn.remotePort}`;
        await conn.stop({ isApp: true, errorCode: CloseReason.ConnectionFromOurself });
        return;
      }

      if (peers.isConnected(lastConnectedPeer.info.id)) {
        logger.log`🛜 Rejecting duplicate connection with peer ${lastConnectedPeer.info.id} from ${conn.remoteHost}:${conn.remotePort}`;
        await conn.stop({ isApp: true, errorCode: CloseReason.DuplicateConnection });
        return;
      }

      logger.log`🛜 Server handshake with ${conn.remoteHost}:${conn.remotePort}`;
      newPeer(conn, lastConnectedPeer.info);
      lastConnectedPeer.info = null;
      await conn.start();
    });

    // connecting to a peer
    async function dial(peer: PeerAddress, options: DialOptions): Promise<QuicPeer> {
      return doDial();

      async function doDial() {
        const peerDetails = peerVerification();
        const clientLater = QUICClient.createQUICClient(
          {
            socket: socket,
            host: peer.host,
            port: peer.port,
            crypto: getQuicClientCrypto(),
            config: {
              ...config,
              verifyCallback: peerDetails.verifyCallback,
            },
            logger: quicLogger.getChild("client"),
          },
          {
            signal: options.signal,
          },
        );
        const client = await clientLater;

        if (peerDetails.info === null) {
          await client.destroy({ isApp: true, errorCode: CloseReason.PeerIdMismatch });
          throw new Error("Client connected, but there is no peer details!");
        }

        if (options.verifyName !== undefined && options.verifyName !== peerDetails.info.id) {
          await client.destroy({ isApp: true, errorCode: CloseReason.PeerIdMismatch });
          throw new Error(
            `Client connected, but the id didn't match. Expected: ${options.verifyName}, got: ${peerDetails.info.id}`,
          );
        }

        addEventListener(client, events.EventQUICClientClose, () => {
          logger.log`⚰️ Client connection closed.`;
        });

        addEventListener(client, events.EventQUICClientError, (error) => {
          logger.error`🔴 Client error: ${error.detail}`;
        });

        logger.log`🤝 Client handshake with: ${peer.host}:${peer.port}`;
        return newPeer(client.connection, peerDetails.info);
      }
    }

    function newPeer(conn: QUICConnection, peerInfo: PeerInfo) {
      const peer = new QuicPeer(conn, peerInfo);
      addEventListener(peer.conn, events.EventQUICConnectionClose, () => peers.peerDisconnected(peer));
      peers.peerConnected(peer);
      return peer;
    }

    return new QuicNetwork(socket, server, dial, peers, { host, port });
  }
}
