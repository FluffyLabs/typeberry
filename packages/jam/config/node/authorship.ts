import type { TimeSlot, ValidatorIndex } from "@typeberry/block";
import { type JsonObject, fromJson } from "@typeberry/block-json";
import type { Bytes } from "@typeberry/bytes";
import type { HASH_SIZE } from "@typeberry/hash";
import { json } from "@typeberry/json-parser";

/** Block authorship options. */
export class AuthorshipOptions {
  static fromJson = json.object<JsonObject<AuthorshipOptions>, AuthorshipOptions>(
    {
      omit_seal_verification: "boolean",
      bandersnatch_seed: json.optional(fromJson.bytes32()),
      bls_seed: json.optional(fromJson.bytes32()),
      ed25519_seed: json.optional(fromJson.bytes32()),
      time_slot: json.optional("number"),
      validator_index: json.optional("number"),
    },
    AuthorshipOptions.new,
  );

  static new({
    omit_seal_verification,
    bandersnatch_seed,
    bls_seed,
    ed25519_seed,
    time_slot,
    validator_index,
  }: JsonObject<AuthorshipOptions>) {
    return new AuthorshipOptions(
      omit_seal_verification,
      bandersnatch_seed,
      bls_seed,
      ed25519_seed,
      time_slot,
      validator_index,
    );
  }

  private constructor(
    /** Use fake seal verification instead of running bandersnatch. */
    public readonly omitSealVerification: boolean,
    /** Use predefined bandersnatch seed to derive bandersnatch key. */
    public readonly bandersnatchSeed?: Bytes<HASH_SIZE>,
    /** Use predefined bls seed to derive bls key. */
    public readonly blsSeed?: Bytes<HASH_SIZE>,
    /** Use predefined ed25519 seed to derive ed25519 key. */
    public readonly ed25519Seed?: Bytes<HASH_SIZE>,
    /** Use to override genesis head config slot. */
    public readonly timeSlot?: TimeSlot,
    /** Use to specify validator index that will be used as for current node. */
    public readonly validatorIndex?: ValidatorIndex,
  ) {}
}
