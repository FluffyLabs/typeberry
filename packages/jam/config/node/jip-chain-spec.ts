import { type JsonObject, fromJson } from "@typeberry/block-json";
import { type Bytes, BytesBlob } from "@typeberry/bytes";
import { Bootnode } from "@typeberry/jamnp-s";
import { json } from "@typeberry/json-parser";
import { isU16 } from "@typeberry/numbers";
import { WithDebug, asOpaqueType } from "@typeberry/utils";

export function parseBootnode(v: string): Bootnode {
  const [name, ipPort] = v.split("@");
  // spliting only by last `:` in case of IPv6
  const ip = ipPort.substring(0, ipPort.lastIndexOf(":"));
  const port = ipPort.substring(ipPort.lastIndexOf(":") + 1);
  if (name === "" || ip === "" || port === "") {
    throw new Error(`Invalid bootnode format, expected: <name>@<ip>:<port>, got: "${v}"`);
  }

  const portNumber = Number.parseInt(port);
  if (!isU16(portNumber)) {
    throw new Error(`Invalid port number: "${port}"`);
  }

  // TODO [ToDr] we should probably validate the name!
  return new Bootnode(asOpaqueType(name), ip, portNumber);
}

/**
 * Configuration file for any blockchain network built on the JAM protocol
 *
 * Provides all the essential information that a new node
 * needs to join, connect to others, and understand the blockchain's starting point
 *
 *  https://github.com/polkadot-fellows/JIPs/blob/90f809b84a9913a821437225f085cf5153870212/JIP-4.md#jip-4-chainspec-file
 */
export class JipChainSpec extends WithDebug {
  /** Optional list of initial contact points for a new node joining the network */
  readonly bootnodes?: Bootnode[];
  /** Human-readable identifier for the network */
  readonly id: string;
  /** Contains the serialized header of the first `block 0` */
  readonly genesisHeader: BytesBlob;
  /**
   * Initial state of the blockchain at the moment of its creation
   *
   * - `key`: is 31-byte object identifier
   * - `value`: is the data stored at this location, as BytesBlob of any length.
   */
  readonly genesisState: Map<Bytes<31>, BytesBlob>;

  static fromJson = json.object<JsonObject<JipChainSpec>, JipChainSpec>(
    {
      bootnodes: json.optional(json.array(json.fromString(parseBootnode))),
      id: "string",
      genesis_header: fromJson.bytesBlobNoPrefix,
      genesis_state: json.map(fromJson.bytesNNoPrefix(31), fromJson.bytesBlobNoPrefix),
    },
    (o) =>
      JipChainSpec.create({
        bootnodes: o.bootnodes,
        id: o.id,
        genesisHeader: o.genesis_header,
        genesisState: o.genesis_state ?? new Map(),
      }),
  );

  static create({
    bootnodes = [],
    id = "",
    genesisHeader = BytesBlob.empty(),
    genesisState = new Map(),
  }: {
    bootnodes?: Bootnode[];
    id?: string;
    genesisHeader?: BytesBlob;
    genesisState?: Map<Bytes<31>, BytesBlob>;
  }) {
    return new JipChainSpec({ bootnodes, id, genesisHeader, genesisState });
  }

  private constructor(data: Pick<JipChainSpec, keyof JipChainSpec>) {
    super();

    this.bootnodes = data.bootnodes;
    this.id = data.id;
    this.genesisHeader = data.genesisHeader;
    this.genesisState = data.genesisState;
  }
}
