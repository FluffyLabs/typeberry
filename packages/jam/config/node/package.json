{
  "name": "@typeberry/config-node",
  "version": "0.0.1",
  "description": "Config for typeberry node.",
  "main": "index.ts",
  "type": "module",
  "scripts": {
    "test": "tsx --test $(find . -type f -name '*.test.ts' | tr '\\n' ' ')"
  },
  "dependencies": {
    "@typeberry/block-json": "0.0.1",
    "@typeberry/bytes": "0.0.1",
    "@typeberry/config": "^0.0.1",
    "@typeberry/configs": "^0.0.1",
    "@typeberry/jamnp-s": "0.0.1",
    "@typeberry/json-parser": "0.0.1",
    "@typeberry/numbers": "0.0.1",
    "@typeberry/utils": "0.0.1"
  },
  "author": "Fluffy Labs",
  "license": "MPL-2.0"
}
