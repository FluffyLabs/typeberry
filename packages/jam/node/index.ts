export * from "./common.js";
export * from "./jam-config.js";
export * from "./main.js";
export * from "./main-fuzz.js";
export { importBlocks } from "./reader.js";
