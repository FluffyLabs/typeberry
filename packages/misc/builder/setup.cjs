const fs = require("node:fs");
const childProcess = require("node:child_process");

const packageToBuild = process.env.PACKAGE_NAME;
let outDir = process.env.PACKAGE_OUT;

const isRelease = process.env.IS_RELEASE !== undefined;

if (!packageToBuild) {
  throw new Error(`Missing 'PACKAGE_NAME' environment variable.`);
}

if (!outDir) {
  const parts = packageToBuild.split("/");
  outDir = parts[parts.length - 1];
}

const DIST = `${__dirname}/../../../dist/${outDir}`;
const inputFile = `${__dirname}/pkg.ts`;

const data = `export * from "${packageToBuild}";`;
fs.writeFileSync(inputFile, data);

const commitHashResult = childProcess.execSync("git rev-parse --short HEAD");
const originalPackageJson = require(`${packageToBuild}/package.json`);
// generate package.json
const packageJson = JSON.stringify(
  {
    name: packageToBuild,
    version: isRelease
      ? originalPackageJson.version
      : `${originalPackageJson.version}-${commitHashResult.toString("utf8").trim()}`,
    main: "index.js",
    author: originalPackageJson.author,
    license: originalPackageJson.license,
    sideEffects: false,
    exports: {
      ".": {
        import: "./index.js",
        require: "./index.cjs",
      },
    },
  },
  null,
  2,
);
fs.mkdirSync(DIST, { recursive: true });
fs.writeFileSync(`${DIST}/package.json`, packageJson);

fs.writeFileSync(
  `${DIST}/.npmignore`,
  `
bin/**
packages/**
`,
);

module.exports = {
  inputFile,
  esmOutFile: `${DIST}/index.js`,
  cjsOutFile: `${DIST}/index.cjs`,
  typesInput: `${DIST}/packages/misc/builder/pkg.d.ts`,
};
