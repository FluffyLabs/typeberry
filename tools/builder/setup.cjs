const fs = require("node:fs");
const childProcess = require("node:child_process");

const packageToBuild = process.env.PACKAGE_NAME;
let outDir = process.env.PACKAGE_OUT;

if (!packageToBuild) {
  throw new Error(`Missing 'PACKAGE_NAME' environment variable.`);
}

if (!outDir) {
  const parts = packageToBuild.split("/");
  outDir = parts[parts.length - 1];
}

const data = `export * from "${packageToBuild}";`;
fs.writeFileSync(`${__dirname}/pkg.ts`, data);

const DIST = `${__dirname}/../../dist/${outDir}`;

const commitHashResult = childProcess.execSync("git rev-parse --short HEAD");
const originalPackageJson = require(`${packageToBuild}/package.json`);
// generate package.json
const packageJson = JSON.stringify(
  {
    name: packageToBuild,
    version: `${originalPackageJson.version}-${commitHashResult.toString("utf8").trim()}`,
    main: "index.js",
    author: originalPackageJson.author,
    license: originalPackageJson.license,
  },
  null,
  2,
);
fs.mkdirSync(DIST, { recursive: true });
fs.writeFileSync(`${DIST}/package.json`, packageJson);

fs.writeFileSync(
  `${DIST}/.npmignore`,
  `tools/**
packages/**
`,
);

module.exports = {
  outFile: `${DIST}/index.js`,
  typesInput: `${DIST}/tools/builder/pkg.d.ts`,
};
