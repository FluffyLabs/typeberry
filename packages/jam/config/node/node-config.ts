import type { JsonObject } from "@typeberry/block-json";
import { type FromJson, json } from "@typeberry/json-parser";
import { JipChainSpec } from "./jip-chain-spec.js";

/** Chain spec chooser. */
export enum KnownChainSpec {
  /** Tiny chain spec. */
  Tiny = "tiny",
  /** Full chain spec. */
  Full = "full",
}

export const knownChainSpecFromJson = json.fromString((input, ctx): KnownChainSpec => {
  switch (input) {
    case KnownChainSpec.Tiny:
      return KnownChainSpec.Tiny;
    case KnownChainSpec.Full:
      return KnownChainSpec.Full;
    default:
      throw Error(`unknown network flavor: ${input} at ${ctx}`);
  }
}) as FromJson<KnownChainSpec>;

export class NodeConfiguration {
  static fromJson = json.object<JsonObject<NodeConfiguration>, NodeConfiguration>(
    {
      $schema: "string",
      version: "number",
      flavor: knownChainSpecFromJson,
      chain_spec: JipChainSpec.fromJson,
      database_base_path: "string",
    },
    NodeConfiguration.new,
  );

  static new({ $schema, version, flavor, chain_spec, database_base_path }: JsonObject<NodeConfiguration>) {
    if (version !== 1) {
      throw new Error("Only version=1 config is supported.");
    }
    return new NodeConfiguration($schema, version, flavor, chain_spec, database_base_path);
  }

  private constructor(
    public readonly $schema: string,
    public readonly version: number,
    public readonly flavor: KnownChainSpec,
    public readonly chainSpec: JipChainSpec,
    public readonly databaseBasePath: string,
  ) {}
}
