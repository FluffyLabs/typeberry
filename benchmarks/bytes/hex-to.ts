import { pathToFileURL } from "node:url";
import { add, complete, configure, cycle, save, suite } from "@typeberry/benchmark/setup.js";

const CODE_OF_0 = "0".charCodeAt(0);
const CODE_OF_a = "a".charCodeAt(0);

const size = 256;
const data: number[] = [];
for (let i = 0; i < size; i += 1) {
  data.push(i);
}

function byteToHexString(byte: number): string {
  const nibbleToString = (n: number) => {
    if (n > 9) {
      return String.fromCharCode(n + CODE_OF_a - 10);
    }
    return String.fromCharCode(n + CODE_OF_0);
  };

  return `${nibbleToString(byte >>> 4)}${nibbleToString(byte & 0xf)}`;
}

export default function run() {
  suite(
    "Bytes / into hex",

    add("number toString + padding", () => {
      return data.map((byte) => byte.toString(16).padStart(2, "0"));
    }),

    add("manual", () => {
      return data.map((byte) => byteToHexString(byte));
    }),

    cycle(),
    complete(),
    configure({}),
    ...save(import.meta.filename),
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run();
}
