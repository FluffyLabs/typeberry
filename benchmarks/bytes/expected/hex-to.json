{
  "name": "Bytes / into hex",
  "date": "2024-09-16T11:15:49.606Z",
  "version": null,
  "results": [
    {
      "name": "number toString + padding",
      "ops": 318696.03,
      "margin": 0.1
    },
    {
      "name": "manual",
      "ops": 358682.53,
      "margin": 0.1
    }
  ],
  "fastest": {
    "name": "manual",
    "index": 1
  },
  "slowest": {
    "name": "number toString + padding",
    "index": 0
  }
}
